### ISSUE TITLE
Invalid generated IL code for null-conditional operator involving a generic parameter

### ISSUE THREAD

**Issue by manu-silicon** [dotnet/roslyn#7388]
If you have:

```
public abstract class A <T> : object where T : class {
    public T Value {get; set; }
    public static explicit operator T (A<T> val) {
        return val?.Value;
    }
}
```

it will generate a DLL for which peverify complains with:

[IL]: Error: [a.dll : A`1[T]::op_Explicit][offset 0x0000000D][found Nullobjref 'NullReference'][expected (unboxed) 'T'] Unexpected type on the stack.
[IL]: Error: [a.dll : A`1[T]::op_Explicit][offset 0x00000008] Non-compatible types on stack depending on path.
2 Error(s) Verifying a.dll

Changing the code to:

```
    public static explicit operator T (A<T> val) {
        return (ReferenceEquals(val,null) ? null : val.Value);
    }
```

will fix the issue.

The problem is that C# emits ldnull when it should use initobj to set the return value to the default type of T.

**Comment by manu-silicon** [comment]
This was using Microsoft (R) Visual C# Compiler version 1.1.0.51109

**Comment by leppie** [comment]
I think this might be false positive of PEVerify. Doing `initobj` does not seem correct to me. The code will run just fine BTW.

**Comment by manu-silicon** [comment]
PEVerify seems correct per ECMA. However I agree that the ECMA rules should be updated to say that ldnull is compatible with T when T has a class constraint, i.e. only a reference type (Currently it says no such thing, it just says T can only be compatible with T).

The reason I posted this is that I encountered an implementation that was crashing when processing this assembly.

**Comment by leppie** [comment]
@manu-silicon That is interesting that it crashed.

**Comment by VSadov** [comment]
null would work at run time, but technically this is not verifiable per verification rules. 
Constraints are mostly considered by verification only when limiting types with which a generic type could be constructed.

**Comment by manu-silicon** [comment]
Thanks for fixing this!
